src/java.base/share/classes/java/util/zip/DeflaterOutputStream.java

```diff
@@ -243,11 +243,12 @@ public void close() throws IOException {
             try {
                 finish();
             } finally {
-                if (usesDefaultDeflater)
+                if (usesDefaultDeflater) {
                     def.end();
+                }
+                out.close();
```

Comment:
This call has a potential to throw an `IOException` in which case any original `IOException` thrown from the `finish()` call will be lost.

Perhaps we should do something like:

```java
public void close() throws IOException {
        if (!closed) {
            IOException finishFailure = null;
            try {
                finish();
            } catch (IOException ioe){
                finishFailure = ioe;
            } finally {
                if (usesDefaultDeflater) {
                    def.end();
                }
                try {
                    out.close();
                } catch (IOException ioe) {
                    if (finishFailure != null) {
                        ioe.addSuppressed(finishFailure);
                    }
                    throw ioe;
                }
                closed = true;
            }
        }
    }

```

Reply:
> This call has a potential to throw an `IOException` in which case any original `IOException` thrown from the `finish()` call will be lost.

Good catch. Adopted your suggestion with a few changes:

- The catch for IOException during finish() needs to rethrow the exception
- The catch for IOException during close() needs to check that finishException != ioe, otherwise addSuppressed may throw IAE, see https://bugs.openjdk.org/browse/JDK-8042377
- I renamed finishFailure to finishException

The above changes were inspired by the overridden method `FilterOutputStream.close()`